{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "48b94d64",
   "metadata": {},
   "source": [
    "# Feature Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a5d658",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dinucleotide-based Cross Covariance (DCC)\n",
    "\n",
    "import os\n",
    "bil='python iLearn-nucleotide-acc.py --file examples/train.txt --method DCC --type DNA --lag 5 --format csv --out DCC.csv'\n",
    "os.system(bil)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1d5ac48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51ce54a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pseudo k-tupler Composition (PseKNC) \n",
    "\n",
    "import os\n",
    "bil='python iLearn-nucleotide-Pse.py --file examples/train.txt --method PseKNC --type DNA --lamada 2  --weight 0.1 --format csv --out PseKNC.csv'\n",
    "os.system(bil)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d32d68ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed28fdfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Series Correlation Pseudo Dinucleotide Composition (SCPseDNC)\n",
    "\n",
    "import os\n",
    "bil='python iLearn-nucleotide-Pse.py --file examples/train.txt --method SCPseDNC --type DNA --lamada 2  --weight 0.1 --format csv --out SCPseDNC.csv'\n",
    "os.system(bil)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2470ba88",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "550d225a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parallel Correlation Pseudo Dinucleotide Composition (PCPseDNC)\n",
    "import os\n",
    "bil='python iLearn-nucleotide-Pse.py --file examples/train.txt --method PCPseDNC --type DNA --lamada 2  --weight 0.1 --format csv --out PCPseDNC.csv'\n",
    "os.system(bil)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aec45693",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46bfbbda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kmer\n",
    "import os\n",
    "bil='python iLearn-nucleotide-basic.py --file examples/train.txt --method Kmer --format csv --out Kmer.csv'\n",
    "os.system(bil)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9e6589c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10dad23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "492da53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Multisource biological features (NPCP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f456929d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os,re, platform\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "import argparse\n",
    "from sklearn.ensemble import AdaBoostClassifier\n",
    " \n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "744a084b",
   "metadata": {},
   "outputs": [],
   "source": [
    "DNAelements = 'ACGT'\n",
    "\n",
    "\n",
    "elements = DNAelements\n",
    "\n",
    "m2 = list(itertools.product(elements, repeat=2))\n",
    "m3 = list(itertools.product(elements, repeat=3))\n",
    "m4 = list(itertools.product(elements, repeat=4))\n",
    "m5 = list(itertools.product(elements, repeat=5))\n",
    "\n",
    "def readFASTAs(filename):\n",
    "    with open(filename, 'r') as file:\n",
    "        v = []\n",
    "        genome = ''\n",
    "        for line in file:\n",
    "            if line[0] != '>':\n",
    "                genome += line.strip()\n",
    "            else:\n",
    "                v.append(genome.upper())\n",
    "                genome = ''\n",
    "        v.append(genome.upper())\n",
    "        del v[0]\n",
    "        return v\n",
    "filename = readFASTAs(\"/Path/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61cca5f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "DNAelements = 'ACGT'\n",
    "\n",
    "\n",
    "elements = DNAelements\n",
    "\n",
    "m2 = list(itertools.product(elements, repeat=2))\n",
    "m3 = list(itertools.product(elements, repeat=3))\n",
    "m4 = list(itertools.product(elements, repeat=4))\n",
    "m5 = list(itertools.product(elements, repeat=5))\n",
    "\n",
    "def readFASTAs(filename):\n",
    "    with open(filename, 'r') as file:\n",
    "        v = []\n",
    "        genome = ''\n",
    "        for line in file:\n",
    "            if line[0] != '>':\n",
    "                genome += line.strip()\n",
    "            else:\n",
    "                v.append(genome.upper())\n",
    "                genome = ''\n",
    "        v.append(genome.upper())\n",
    "        del v[0]\n",
    "        return v\n",
    "\n",
    "def saveCSV(name, X, Y, type):\n",
    "    if type == 'test':\n",
    "        F = open((name + '_testDataset.csv'), 'w')\n",
    "\n",
    "    else:\n",
    "        if type == 'optimum':\n",
    "            F = open((name + '_optimumDataset.csv'), 'w')\n",
    "\n",
    "    for x, y in zip(X, Y):\n",
    "        for each in x:\n",
    "            F.write(str(each) + ',')\n",
    "        F.write(str(int(y)) + '\\n')\n",
    "    F.close()\n",
    "\n",
    "def readLabels(fileName):\n",
    "    with open(fileName, 'r') as file:\n",
    "        v = []\n",
    "        for line in file:\n",
    "            if line != '\\n':\n",
    "                v.append((line.replace('\\n', '')).replace(' ', ''))\n",
    "        return v\n",
    "\n",
    "def kmers(seq, k):\n",
    "    v = []\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        v.append(seq[i:i + k])\n",
    "    return v\n",
    "\n",
    "def pseudoKNC(x, k):\n",
    "    t = []\n",
    "    for i in range(1, k + 1, 1):\n",
    "        v = list(itertools.product(elements, repeat=i))\n",
    "        for i in v:\n",
    "            t.append(x.count(''.join(i)))\n",
    "    return t\n",
    "\n",
    "def zCurve(x):\n",
    "    t = []\n",
    "    T = x.count('T'); A = x.count('A'); C = x.count('C'); G = x.count('G');\n",
    "    x_ = (A + G) - (C + T)\n",
    "    y_ = (A + C) - (G + T)\n",
    "    z_ = (A + T) - (C + G)\n",
    "    t.append(x_); t.append(y_); t.append(z_)\n",
    "    return t\n",
    "\n",
    "def gcContent(x):\n",
    "    t = []\n",
    "    T = x.count('T')\n",
    "    A = x.count('A');\n",
    "    C = x.count('C');\n",
    "    G = x.count('G');\n",
    "    t.append( (G + C) / (A + C + G + T)  * 100.0 )\n",
    "    return t\n",
    "\n",
    "\n",
    "def cumulativeSkew(x):\n",
    "    \n",
    "    t = []\n",
    "    T = x.count('T')\n",
    "    A = x.count('A');\n",
    "    C = x.count('C');\n",
    "    G = x.count('G');\n",
    "\n",
    "    GCSkew = (G-C)/(G+C)\n",
    "    ATSkew = (A-T)/(A+T)\n",
    "\n",
    "    t.append(GCSkew)\n",
    "    t.append(ATSkew)\n",
    "    return t\n",
    "\n",
    "\n",
    "def atgcRatio(x):\n",
    "    t = []\n",
    "    T = x.count('T')\n",
    "    A = x.count('A');\n",
    "    C = x.count('C');\n",
    "    G = x.count('G');\n",
    "\n",
    "    t.append( (A+T)/(G+C) )\n",
    "    return t\n",
    "\n",
    "\n",
    "def monoMonoKGap(x, g):  # k=1, m=1\n",
    "    m = m2\n",
    "    t = []\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 2)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[-1] == gGap[1]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "\n",
    "def monoDiKGap(x, g):  # k=1, m=2\n",
    "    t = []\n",
    "    m = m3\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 3)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[-2] == gGap[1] and v[-1] == gGap[2]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "\n",
    "def diMonoKGap(x, g):  # k=2, m=1\n",
    "    t = []\n",
    "    m = m3\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 3)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[1] == gGap[1] and v[-1] == gGap[2]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "def monoTriKGap(x, g):  # k=1, m=3\n",
    "    t = []\n",
    "    m = m4\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 4)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[-3] == gGap[1] and v[-2] == gGap[2] and v[-1] == gGap[3]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "def triMonoKGap(x, g):  # k=3, m=1\n",
    "    t = []\n",
    "    m = m4\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 4)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[1] == gGap[1] and v[2] == gGap[2] and v[-1] == gGap[3]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "def diDiKGap(x, g): # k=2, m=2\n",
    "    m = m4\n",
    "    t = []\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 4)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[1] == gGap[1] and v[-2] == gGap[2] and v[-1] == gGap[3]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "def diTriKGap(x, g):  # k=2, m=3\n",
    "    m = m5\n",
    "    t = []\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 5)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[1] == gGap[1] and v[-3] == gGap[2] and v[-2] == gGap[3] and v[-1] == gGap[4]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "def triDiKGap(x, g):  # k=3, m=2\n",
    "    m = m5\n",
    "    t = []\n",
    "    for i in range(1, g + 1, 1):\n",
    "        V = kmers(x, i + 5)\n",
    "        for gGap in m:\n",
    "            C = 0\n",
    "            for v in V:\n",
    "                if v[0] == gGap[0] and v[1] == gGap[1] and v[2] == gGap[2] and v[-2] == gGap[3] and v[-1] == gGap[4]:\n",
    "                    C += 1\n",
    "            t.append(C)\n",
    "    return t\n",
    "def generateFeatures(kGap, kTuple, x, y):\n",
    "        feature = []\n",
    "        t = zCurve(x)         \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t= gcContent(x)           \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = cumulativeSkew(x)      \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = atgcRatio(x)         \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = pseudoKNC(x, kTuple)            \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = monoMonoKGap(x, kGap)      \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = monoDiKGap(x, kGap)        \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = monoTriKGap(x, kGap)       \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = diMonoKGap(x, kGap)        \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = diDiKGap(x, kGap)          \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = diTriKGap(x, kGap)        \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = triMonoKGap(x, kGap)       \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        t = triDiKGap(x, kGap)         \n",
    "        for item in t:\n",
    "            feature.append(item)\n",
    "        feature.append(y)\n",
    "        return feature\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "408b1d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate Train data Features\n",
    "X = readFASTAs(\"/Path/\")\n",
    "print('loading sequence complete')\n",
    "Y = readLabels(\"/Path/\")\n",
    "print('loading label complete')\n",
    "Y = LabelEncoder().fit_transform(Y)\n",
    "assert len(X)==len(Y)\n",
    "T = []\n",
    "print('start generating features')\n",
    "for x, y in zip(X, Y):\n",
    "    \n",
    "    feature = generateFeatures(5, 3, x, y)\n",
    "    print(len(feature))\n",
    "    T.append(feature)\n",
    "    print(len(T))\n",
    "T = np.array(T)\n",
    "\n",
    "X = T[:,:-1]\n",
    "Y = T[:,-1]\n",
    "print('Finish generating features')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27b4a3dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# X = pd.DataFrame(X)\n",
    "# X.to_csv(\"/home/bilal/Biomed/Dataset/ilearn/kGAP_kTuple_train.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c5f3220",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(\"/path/name.csv\", X,\n",
    "              delimiter = \",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee49483c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate Independent data features\n",
    "X = readFASTAs(\"/path/\")\n",
    "print('loading sequence complete')\n",
    "Y = readLabels(\"/path/\")\n",
    "print('loading label complete')\n",
    "Y = LabelEncoder().fit_transform(Y)\n",
    "assert len(X)==len(Y)\n",
    "T = []\n",
    "print('start generating features')\n",
    "for x, y in zip(X, Y):\n",
    "    \n",
    "    feature = generateFeatures(5, 3, x, y)\n",
    "    print(len(feature))\n",
    "    T.append(feature)\n",
    "    print(len(T))\n",
    "T = np.array(T)\n",
    "\n",
    "X = T[:,:-1]\n",
    "Y = T[:,-1]\n",
    "print('Finish generating features')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a61d99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# X = pd.DataFrame(X)\n",
    "# X.to_csv(\"/home/bilal/Biomed/Dataset/ilearn/kGAP_kTuple_ind.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c62b0d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(\"/path/name.csv\", X,\n",
    "              delimiter = \",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbf7b850",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5382ac4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d56adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "\n",
    "# Load the data\n",
    "data = pd.read_csv(\"/data/\", header = None, names = ['sequence'])\n",
    "\n",
    "# Clean the data\n",
    "data = data.dropna()\n",
    "\n",
    "# Extract features\n",
    "vectorizer = CountVectorizer(ngram_range=(3,3), analyzer='char')\n",
    "X = vectorizer.fit_transform(data['sequence']).toarray()\n",
    "\n",
    "# Scale the features\n",
    "scaler = StandardScaler()\n",
    "X = scaler.fit_transform(X)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c48f9f45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b255a52",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
